File: backend/app/services/execute_service.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal, ROUND_FLOOR
from zoneinfo import ZoneInfo

from sqlalchemy.orm import Session

from app.models.execute_row_model import ExecuteRow
from app.models.script_model import Script
from app.schemas.execute_schema import ExecuteAlert, ExecuteDetailsResponse, ExecuteRowCreate, ExecuteRowResponse, ExecuteRowsResponse
from app.services import trigger_lifecycle
from app.services.telegram_alerts import clear_all_edge_state_for_script, get_trigger_visual_states, notify_row_after_save


IST = ZoneInfo("Asia/Kolkata")


def _utc_to_naive_ist(dt: datetime) -> datetime:
    """updated_at is stored as naive UTC (datetime.utcnow); trade_date/trade_time are naive IST
    wall-clock. Convert to naive IST so both timestamp sources agree in the alerts feed."""
    return dt.replace(tzinfo=ZoneInfo("UTC")).astimezone(IST).replace(tzinfo=None)


def _d(value: object, places: int = 2) -> Decimal:
    if value is None:
        return Decimal("0").quantize(Decimal("1").scaleb(-places))
    dec = value if isinstance(value, Decimal) else Decimal(str(value))
    return dec.quantize(Decimal("1").scaleb(-places))


def _script_or_none(db: Session, script_id: int) -> Script | None:
    return db.query(Script).filter(Script.id == script_id).first()


def _rows(db: Session, script_id: int) -> list[ExecuteRow]:
    return db.query(ExecuteRow).filter(ExecuteRow.script_id == script_id).order_by(ExecuteRow.id.asc()).all()


def _slot_amount(script: Script) -> Decimal:
    # max_buy is intentionally excluded from sizing for now — see the notes on the row-cap
    # removal in add_execute_row / auto_trade_service. Only invest_slots divides the budget.
    denom = int(script.invest_slots)
    if denom <= 0:
        return Decimal("0")
    return _d(script.total_invest_amount, 2) / Decimal(denom)


def _calc_qty(slot_amount: Decimal, buy_price: Decimal) -> Decimal:
    """Whole shares only (NSE equities can't be bought/sold fractionally) — floor to the nearest unit."""
    if buy_price <= 0:
        return Decimal("0")
    return (slot_amount / buy_price).quantize(Decimal("1"), rounding=ROUND_FLOOR)


def row_to_response(db: Session, row: ExecuteRow) -> ExecuteRowResponse | None:
    """Build API row payload including live trigger fields."""
    script = _script_or_none(db, row.script_id)
    if not script:
        return None
    _, _, current_price, _, _, _ = trigger_lifecycle.prepare_script_triggers(db, script)
    all_rows = _rows(db, row.script_id)
    return _serialize_row(script, row, all_rows, current_price)


def _serialize_row(
    script: Script, row: ExecuteRow, all_rows: list[ExecuteRow], current_price: Decimal | None
) -> ExecuteRowResponse:
    up_pct = _d(script.up_percent, 2)
    down_pct = _d(script.down_percent, 2)
    complete = trigger_lifecycle.is_trade_complete(row)
    if complete:
        lbt = lst = None
    else:
        ref = trigger_lifecycle.reference_for_active_row(row, all_rows)
        if ref is not None and ref > 0:
            lbt, lst = trigger_lifecycle.triggers_from_reference(ref, up_pct, down_pct)
        else:
            lbt = lst = None
    vis = get_trigger_visual_states(int(script.id), current_price, lbt, lst)
    return ExecuteRowResponse(
        id=row.id,
        script_id=row.script_id,
        trade_date=row.trade_date,
        trade_time=row.trade_time,
        buy_price=row.buy_price,
        buy_qty=row.buy_qty,
        sell_price=row.sell_price,
        sell_qty=row.sell_qty,
        up_value=row.up_value,
        down_value=row.down_value,
        qty_left=row.qty_left,
        created_at=row.created_at,
        updated_at=_utc_to_naive_ist(row.updated_at),
        is_trade_complete=complete,
        live_buy_trigger=lbt,
        live_sell_trigger=lst,
        up_visual_state=vis["sell_state"],
        down_visual_state=vis["buy_state"],
    )


def get_execute_details(db: Session, script_id: int) -> ExecuteDetailsResponse | None:
    script = _script_or_none(db, script_id)
    if not script:
        return None

    ex_rows, _, current_price, buy_trigger, sell_trigger, _pairs = trigger_lifecycle.prepare_script_triggers(db, script)
    vis = get_trigger_visual_states(int(script.id), current_price, buy_trigger, sell_trigger)

    up_percent = _d(script.up_percent, 2)
    down_percent = _d(script.down_percent, 2)

    # Only currently-open positions count as "invested" — a completed (sold) row's qty_left is 0,
    # so its original buy cost naturally drops out once the trade is booked.
    invested_amount = sum((r.buy_price * r.qty_left for r in ex_rows), start=Decimal("0.00")).quantize(Decimal("0.01"))
    total_quantity = sum((r.qty_left for r in ex_rows), start=Decimal("0.00")).quantize(Decimal("0.01"))
    current_value = (
        (total_quantity * current_price).quantize(Decimal("0.01"))
        if current_price is not None
        else Decimal("0.00")
    )
    current_holdings = current_value
    positions = sum(1 for r in ex_rows if r.qty_left > 0)

    alerts: list[ExecuteAlert] = []
    for r in ex_rows:
        alerts.append(
            ExecuteAlert(
                side="BUY",
                price=_d(r.buy_price, 2),
                timestamp=datetime.combine(r.trade_date, r.trade_time),
            )
        )
        if r.sell_price is not None:
            alerts.append(
                ExecuteAlert(
                    side="SELL",
                    price=_d(r.sell_price, 2),
                    timestamp=_utc_to_naive_ist(r.updated_at),
                )
            )
    alerts.sort(key=lambda a: a.timestamp, reverse=True)

    return ExecuteDetailsResponse(
        script_id=script.id,
        script_name=script.name,
        current_market_price=current_price,
        buy_trigger_value=buy_trigger,
        sell_trigger_value=sell_trigger,
        invested_amount=invested_amount,
        current_value=current_value,
        current_holdings=current_holdings,
        current_holding_positions=positions,
        total_quantity=total_quantity,
        up_percent=up_percent,
        down_percent=down_percent,
        total_invest_amount=_d(script.total_invest_amount, 2),
        invest_slots=int(script.invest_slots),
        max_buy=int(script.max_buy),
        max_rows=int(script.invest_slots) * int(script.max_buy),
        recent_alerts=alerts[:5],
        buy_visual_state=vis["buy_state"],
        sell_visual_state=vis["sell_state"],
    )


def get_execute_rows(db: Session, script_id: int) -> ExecuteRowsResponse | None:
    script = _script_or_none(db, script_id)
    if not script:
        return None
    _, _, current_price, _, _, _ = trigger_lifecycle.prepare_script_triggers(db, script)
    rows = _rows(db, script_id)
    return ExecuteRowsResponse(rows=[_serialize_row(script, r, rows, current_price) for r in rows])


def add_execute_row(db: Session, script_id: int, payload: ExecuteRowCreate, notify: bool = True) -> ExecuteRow | None:
    script = _script_or_none(db, script_id)
    if not script:
        return None

    # Cap concurrent open positions at invest_slots — each slot commits total_invest_amount /
    # invest_slots, so letting more than invest_slots stay open at once overcommits capital
    # beyond total_invest_amount. (max_buy lifetime-row cap is a separate, still-unenforced idea.)
    open_positions = sum(1 for r in _rows(db, script_id) if r.qty_left > 0)
    if open_positions >= int(script.invest_slots):
        raise ValueError("NO_SLOT_AVAILABLE")

    buy_price = _d(payload.buy_price, 2)
    slot_amount = _slot_amount(script)
    buy_qty = _calc_qty(slot_amount, buy_price)
    if buy_qty <= 0:
        raise ValueError("ZERO_QTY")
    sell_price = _d(payload.sell_price, 2) if payload.sell_price is not None else None
    sell_qty = buy_qty if sell_price is not None else None

    up_value = (buy_price * (Decimal("1") + (_d(script.up_percent, 2) / Decimal("100")))).quantize(Decimal("0.01"))
    down_value = (buy_price * (Decimal("1") - (_d(script.down_percent, 2) / Decimal("100")))).quantize(Decimal("0.01"))
    qty_left = Decimal("0.00") if sell_price is not None else buy_qty

    row = ExecuteRow(
        script_id=script_id,
        trade_date=payload.trade_date,
        trade_time=payload.trade_time,
        buy_price=buy_price,
        buy_qty=buy_qty,
        sell_price=sell_price,
        sell_qty=sell_qty,
        up_value=up_value,
        down_value=down_value,
        qty_left=qty_left,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    trigger_lifecycle.rebuild_post_sell_anchor(db, script_id)
    clear_all_edge_state_for_script(script_id)
    if notify and not trigger_lifecycle.is_trade_complete(row):
        notify_row_after_save(db, script, row)
    return row


def update_execute_row(db: Session, row_id: int, payload: ExecuteRowCreate, notify: bool = True) -> ExecuteRow | None:
    row = db.query(ExecuteRow).filter(ExecuteRow.id == row_id).first()
    if not row:
        return None
    script = _script_or_none(db, row.script_id)
    if not script:
        return None

    buy_price = _d(payload.buy_price, 2)
    slot_amount = _slot_amount(script)
    buy_qty = _calc_qty(slot_amount, buy_price)
    sell_price = _d(payload.sell_price, 2) if payload.sell_price is not None else None
    sell_qty = buy_qty if sell_price is not None else None

    up_value = (buy_price * (Decimal("1") + (_d(script.up_percent, 2) / Decimal("100")))).quantize(Decimal("0.01"))
    down_value = (buy_price * (Decimal("1") - (_d(script.down_percent, 2) / Decimal("100")))).quantize(Decimal("0.01"))
    qty_left = Decimal("0.00") if sell_price is not None else buy_qty

    row.trade_date = payload.trade_date
    row.trade_time = payload.trade_time
    row.buy_price = buy_price
    row.buy_qty = buy_qty
    row.sell_price = sell_price
    row.sell_qty = sell_qty
    row.up_value = up_value
    row.down_value = down_value
    row.qty_left = qty_left

    db.commit()
    db.refresh(row)
    trigger_lifecycle.rebuild_post_sell_anchor(db, row.script_id)
    clear_all_edge_state_for_script(row.script_id)
    if notify and not trigger_lifecycle.is_trade_complete(row):
        notify_row_after_save(db, script, row)
    return row

File: backend/supabase/functions/_shared/executeService.ts
import { Decimal, d, decOut, floorInt } from "./decimal.ts";
import { fetchRowById, fetchRowsForScript, insertRow, updateRowById, type ExecuteRow } from "./executeRows.ts";
export type { ExecuteRow } from "./executeRows.ts";
import { getScript, type Script } from "./scriptService.ts";
import * as tl from "./triggerLifecycle.ts";
import { clearAllEdgeStateForScript, getTriggerVisualStates, notifyRowAfterSave } from "./telegramAlerts.ts";
import { utcInstantToNaiveIstString } from "./wallclock.ts";

export class NoSlotAvailableError extends Error {}
export class ZeroQtyError extends Error {}

export type ExecuteRowInput = {
  trade_date: string;
  trade_time: string;
  buy_price: number;
  sell_price: number | null;
};

function slotAmount(script: Script): Decimal {
  const denom = script.invest_slots;
  if (denom <= 0) return new Decimal(0);
  return d(script.total_invest_amount, 2).div(denom);
}

function calcQty(slotAmt: Decimal, buyPrice: Decimal): Decimal {
  if (buyPrice.lte(0)) return new Decimal(0);
  return floorInt(slotAmt.div(buyPrice));
}

export async function serializeRow(
  script: Script,
  row: ExecuteRow,
  allRows: ExecuteRow[],
  currentPrice: Decimal | null,
): Promise<Record<string, unknown>> {
  const upPct = d(script.up_percent, 2);
  const downPct = d(script.down_percent, 2);
  const complete = tl.isTradeComplete(row);
  let lbt: Decimal | null = null;
  let lst: Decimal | null = null;
  if (!complete) {
    const ref = tl.referenceForActiveRow(row, allRows);
    if (ref !== null && ref.gt(0)) {
      [lbt, lst] = tl.triggersFromReference(ref, upPct, downPct);
    }
  }
  const vis = await getTriggerVisualStates(script.id, currentPrice, lbt, lst);
  return {
      id: row.id,
      script_id: row.script_id,
      trade_date: row.trade_date,
      trade_time: row.trade_time,
      buy_price: decOut(row.buy_price),
      buy_qty: decOut(row.buy_qty, 6),
      sell_price: decOut(row.sell_price),
      sell_qty: decOut(row.sell_qty, 6),
      up_value: decOut(row.up_value),
      down_value: decOut(row.down_value),
      qty_left: decOut(row.qty_left, 6),
      created_at: row.created_at.toISOString(),
      updated_at: utcInstantToNaiveIstString(row.updated_at),
      is_trade_complete: complete,
      live_buy_trigger: decOut(lbt),
      live_sell_trigger: decOut(lst),
      // Deliberately crossed, matching the original: up_visual_state <- sell_state, down_visual_state <- buy_state.
    up_visual_state: vis.sell_state,
    down_visual_state: vis.buy_state,
  };
}

export async function getExecuteDetails(scriptId: number): Promise<Record<string, unknown> | null> {
  const script = await getScript(scriptId);
  if (!script) return null;

  const { rows: exRows, current, buyTrigger, sellTrigger } = await tl.prepareScriptTriggers(script);
  const vis = await getTriggerVisualStates(script.id, current, buyTrigger, sellTrigger);

  const upPercent = d(script.up_percent, 2);
  const downPercent = d(script.down_percent, 2);

  // Only currently-open positions count as "invested" — a completed (sold) row's qty_left is 0.
  let investedAmount = new Decimal(0);
  let totalQuantity = new Decimal(0);
  for (const r of exRows) {
    investedAmount = investedAmount.plus(d(r.buy_price).times(d(r.qty_left, 6)));
    totalQuantity = totalQuantity.plus(d(r.qty_left, 6));
  }
  investedAmount = investedAmount.toDecimalPlaces(2);
  totalQuantity = totalQuantity.toDecimalPlaces(2);

  const currentValue =
    current !== null ? totalQuantity.times(current).toDecimalPlaces(2) : new Decimal(0).toDecimalPlaces(2);
  const currentHoldings = currentValue;
  const positions = exRows.filter((r) => d(r.qty_left, 6).gt(0)).length;

  type Alert = { side: string; price: Decimal; timestamp: string };
  const alerts: Alert[] = [];
  for (const r of exRows) {
    alerts.push({ side: "BUY", price: d(r.buy_price, 2), timestamp: `${r.trade_date}T${r.trade_time}` });
    if (r.sell_price !== null) {
      alerts.push({ side: "SELL", price: d(r.sell_price, 2), timestamp: utcInstantToNaiveIstString(r.updated_at) });
    }
  }
  alerts.sort((a, b) => (a.timestamp < b.timestamp ? 1 : a.timestamp > b.timestamp ? -1 : 0));

  return {
    script_id: script.id,
    script_name: script.name,
    current_market_price: decOut(current),
    buy_trigger_value: decOut(buyTrigger),
    sell_trigger_value: decOut(sellTrigger),
    invested_amount: decOut(investedAmount),
    current_value: decOut(currentValue),
    current_holdings: decOut(currentHoldings),
    current_holding_positions: positions,
    total_quantity: decOut(totalQuantity),
    up_percent: decOut(upPercent),
    down_percent: decOut(downPercent),
    total_invest_amount: decOut(script.total_invest_amount),
    invest_slots: script.invest_slots,
    max_buy: script.max_buy,
    max_rows: script.invest_slots * script.max_buy,
    recent_alerts: alerts.slice(0, 5).map((a) => ({ side: a.side, price: decOut(a.price), timestamp: a.timestamp })),
    buy_visual_state: vis.buy_state,
    sell_visual_state: vis.sell_state,
  };
}

export async function getExecuteRows(scriptId: number): Promise<{ rows: Record<string, unknown>[] } | null> {
  const script = await getScript(scriptId);
  if (!script) return null;
  const { current } = await tl.prepareScriptTriggers(script);
  const rows = await fetchRowsForScript(scriptId, "asc");
  return { rows: await Promise.all(rows.map((r) => serializeRow(script, r, rows, current))) };
}

export async function addExecuteRow(
  scriptId: number,
  payload: ExecuteRowInput,
  notify = true,
): Promise<ExecuteRow | null> {
  const script = await getScript(scriptId);
  if (!script) return null;

  // Cap concurrent open positions at invest_slots — each slot commits total_invest_amount /
  // invest_slots, so letting more than invest_slots stay open at once overcommits capital.
  const existing = await fetchRowsForScript(scriptId, "asc");
  const openPositions = existing.filter((r) => d(r.qty_left, 6).gt(0)).length;
  if (openPositions >= script.invest_slots) {
    throw new NoSlotAvailableError();
  }

  const buyPrice = d(payload.buy_price, 2);
  const slotAmt = slotAmount(script);
  const buyQty = calcQty(slotAmt, buyPrice);
  if (buyQty.lte(0)) throw new ZeroQtyError();
  const sellPrice = payload.sell_price !== null ? d(payload.sell_price, 2) : null;
  const sellQty = sellPrice !== null ? buyQty : null;

  const upValue = buyPrice.times(new Decimal(1).plus(d(script.up_percent, 2).div(100))).toDecimalPlaces(2);
  const downValue = buyPrice.times(new Decimal(1).minus(d(script.down_percent, 2).div(100))).toDecimalPlaces(2);
  const qtyLeft = sellPrice !== null ? new Decimal(0).toDecimalPlaces(2) : buyQty;

  const row = await insertRow({
    script_id: scriptId,
    trade_date: payload.trade_date,
    trade_time: payload.trade_time,
    buy_price: buyPrice.toFixed(2),
    buy_qty: buyQty.toFixed(6),
    sell_price: sellPrice !== null ? sellPrice.toFixed(2) : null,
    sell_qty: sellQty !== null ? sellQty.toFixed(6) : null,
    up_value: upValue.toFixed(2),
    down_value: downValue.toFixed(2),
    qty_left: qtyLeft.toFixed(6),
  });

  await tl.rebuildPostSellAnchor(scriptId);
  await clearAllEdgeStateForScript(scriptId);
  if (notify && !tl.isTradeComplete(row)) {
    await notifyRowAfterSave(script, row);
  }
  return row;
}

export async function updateExecuteRow(
  rowId: number,
  payload: ExecuteRowInput,
  notify = true,
): Promise<ExecuteRow | null> {
  const existingRow = await fetchRowById(rowId);
  if (!existingRow) return null;
  const script = await getScript(existingRow.script_id);
  if (!script) return null;

  const buyPrice = d(payload.buy_price, 2);
  const slotAmt = slotAmount(script);
  const buyQty = calcQty(slotAmt, buyPrice);
  const sellPrice = payload.sell_price !== null ? d(payload.sell_price, 2) : null;
  const sellQty = sellPrice !== null ? buyQty : null;

  const upValue = buyPrice.times(new Decimal(1).plus(d(script.up_percent, 2).div(100))).toDecimalPlaces(2);
  const downValue = buyPrice.times(new Decimal(1).minus(d(script.down_percent, 2).div(100))).toDecimalPlaces(2);
  const qtyLeft = sellPrice !== null ? new Decimal(0).toDecimalPlaces(2) : buyQty;

  const row = await updateRowById(rowId, {
    trade_date: payload.trade_date,
    trade_time: payload.trade_time,
    buy_price: buyPrice.toFixed(2),
    buy_qty: buyQty.toFixed(6),
    sell_price: sellPrice !== null ? sellPrice.toFixed(2) : null,
    sell_qty: sellQty !== null ? sellQty.toFixed(6) : null,
    up_value: upValue.toFixed(2),
    down_value: downValue.toFixed(2),
    qty_left: qtyLeft.toFixed(6),
  });
  if (!row) return null;

  await tl.rebuildPostSellAnchor(row.script_id);
  await clearAllEdgeStateForScript(row.script_id);
  if (notify && !tl.isTradeComplete(row)) {
    await notifyRowAfterSave(script, row);
  }
  return row;
}
